### PR TITLE
feat: add document store

### PR DESCRIPTION
**Pull request by mentos1386**
I'm working on implementation of document store. it's still WIP, will hopefully be done soon. All comments are welcome.

JS implementation: https://github.com/orbitdb/orbit-db-docstore

**Implementation progress:**
- [x] Get
- [x] Put
- [x] Delete
- [x] Query
- [x] BatchPut
- [x] PutAll

(related to : https://github.com/berty/go-orbit-db/issues/9)

### CODE REVIEW

**Comment by glouvigny** [comment]
Hi, thanks for opening this PR, don't hesitate to ask questions if you need help :)

I saw your comment about the indexBy field, I think we need to add a store specific field in iface.NewStoreOptions (probably an interface{} casted by each store relying on it). We can implement it in another PR if you want.

**Comment by mentos1386** [comment]
Due to lack of time, i'm not sure when i'll continue working on this.

@glouvigny can this be reviewed and missing functionality be added later, with different PRs?

File: tests/docs_test.go [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+			c.So(len(docs), ShouldEqual, 1)
+			c.So(docs[0], ShouldResemble, document)
+		})
+
+		c.Convey("get", FailureHalts, func(c C) {
+			document := createDocument("doc1", "hello", "world")
+			_, err := db.Put(ctx, document)
+			c.So(err, ShouldBeNil)
+
+			docs, err := db.Get(ctx, "doc1", true)
```
Comment (glouvigny) [inline tests/docs_test.go:89]: Can you ensure that retrieving entries without case-sensitivity works? Specifically I had issues with a Get() partial match

**Comment by moul** [comment]
@mentos1386 @glouvigny do you plan to make changes soon?

would be nice to merge something shortly to avoid having a standalone branch for too long

what are the minimal amount of changes needed to make this PR useful enough and safe to be merged?

**Comment by mentos1386** [comment]
@moul Sorry for taking such a long time. I'll fix the rest of issues tomorrow.

**Comment by chrispanag** [comment]
Hello all, is there any roadmap for this feature to be released? Do you need any help with the implementation?

**Comment by glouvigny** [comment]
Hi @chrispanag, I just rebased, cleaned the code and added batched put methods.

Can you have a look and tell me if that suits your needs?

**Comment by chrispanag** [comment]
Thank you very much @glouvigny! This is amazing work. 

For our use-case the query method would be necessary as well. If you don't have the time to implement it, I could try :)

**Comment by chrispanag** [comment]
@glouvigny Again thank you really much! Your work is greatly appreciated! 💯

**Pull request merged**